codalab/client/json_api_client.py

```diff
@@ -713,3 +713,25 @@ def get_bundle_locations(self, bundle_uuid):
             method='GET', path='/bundles/{}/locations/'.format(bundle_uuid),
         )
         return response['data']
+
+    @wrap_exception('Unable to create the location of bundles')
+    def add_bundle_location(self, bundle_uuid, bundle_store_uuid, params):
+        response = self._make_request(
+            method='POST',
+            path='/bundles/{}/locations/'.format(bundle_uuid),
+            data=self._pack_document(
+                [{'bundle_uuid': bundle_uuid, 'bundle_store_uuid': bundle_store_uuid}],
+                'bundle_locations',
+            ),
+            query_params=self._pack_params(params),
+        )
+        return response['data']
+
+    @wrap_exception("Unable to finalize the status of blob storage bundles")
+    def update_bundle_location_blob(self, bundle_uuid, params):
```

Comment:
```suggestion
    def update_bundle_location(self, bundle_uuid, location_id, params):
```

Reply:
The problem is: not every bundle has a corresponding location_id. For example, if a user does not specify destination storage, the rest server supports using Azure as default storage.  In this scenario, the rest server will pick the default azure storage to store it, and will not add a line in the `bundle_location` table.

Now, the behavior is same for bypass upload and go-through server upload. If we want to add a location in `bundle_location` table for every bundle (and stored in any bundle storage), should we make another PR to change bypass server upload and go-through server upload at the same time?

Reply:
And for now, this function `update_bundle_location` does not really update the bundle location. It only update the bundle state  (success or failure), to finalize the bypass server upload. 
So I think a better way is to change the function name to `update_bundle_state`, and change endpoint URL to `'/bundles/<bundle_uuid:re:%s>/state'`.

Reply:
> The problem is: not every bundle has a corresponding location_id. For example, if a user does not specify destination storage, the rest server supports using Azure as default storage. In this scenario, the rest server will pick the default azure storage to store it, and will not add a line in the `bundle_location` table.

@wwwjn If the rest server sets the CODALAB_ALWAYS_USE_AZURE_BLOB_BETA variable, then we should actually not support bypass server for uploads. This is because, just like cl upload -a, CODALAB_ALWAYS_USE_AZURE_BLOB_BETA is also a temporary / soon-to-be-deprecated solution because it uses storage_type instead of creating an entry in the bundle location table.

This means that we should only support bypass server for uploads where we do create a bundle store, so this means that each bundle here should have a corresponding location_id. This means that there are only two scenarios in which this would be the case:
1. User specifies --store [store name]
2. Server is configured with CODALAB_DEFAULT_BUNDLE_STORE_NAME=[store name]

Does that make sense? I've also added suggestions to the doc (https://docs.google.com/document/d/1qFT6OWSUW5-2ZfnkK_q6P-Q-9sVe_fyBLX3UfEIWRiY/edit?pli=1#) to reflect this change:

<img width="670" alt="image" src="https://user-images.githubusercontent.com/1689183/174839038-aac51e61-bf5c-40a8-9fd8-920c1d1d7fb7.png">

Reply:
I think this is a good solution for this. But I'm a little bit confused about do we need to change this in this PR? Or we simply do not support CODALAB_ALWAYS_USE_BLOB_STORAGE?

If we want to make change for default storage, maybe we need another PR to change go through server at same time

Reply:
I think we should not support bypassing server upload for CODALAB_ALWAYS_USE_AZURE_BLOB_BETA. I will indicate in another comment where I think you need to change this PR.